### PR TITLE
Optimize memory allocations in the event processing pipeline

### PR DESCRIPTION
**Pull request by rdner**
Previously, every time a processor ran on an event we made a clone of the entire event for two reasons:

1. this event could have some nested maps that are shared among multiple events.

2. in case a processor fails to make a change it should be able to revert its partial changes.

This change added a new `EventEditor` wrapper that is used for collecting pending event changes in processors with an option to `Apply` or `Reset` them.

Additionally, this `EventEditor` takes care of the efficient memory management when making changes to an event by cloning only the nested maps that processors access or modify. Most of the processors just put new keys or delete existing keys on the root-level, so most of the time the nested maps in the event remain untouched and it does not require the whole event to be cloned.

Processors migration progress:
- [ ] [add_cloudfoundry_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-cloudfoundry-metadata.html)
- [x] [add_docker_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-docker-metadata.html)
- [x] [add_host_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-host-metadata.html)
- [x] [add_id](https://www.elastic.co/guide/en/beats/filebeat/current/add-id.html)
- [ ] [add_kubernetes_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-kubernetes-metadata.html)
- [ ] [add_locale](https://www.elastic.co/guide/en/beats/filebeat/current/add-locale.html)
- [ ] [add_nomad_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-nomad-metadata.html)
- [ ] [add_observer_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-observer-metadata.html)
- [ ] [add_process_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-process-metadata.html)
- [ ] [community_id](https://www.elastic.co/guide/en/beats/filebeat/current/community-id.html)
- [ ] [convert](https://www.elastic.co/guide/en/beats/filebeat/current/convert.html)
- [ ] [decode_cef](https://www.elastic.co/guide/en/beats/filebeat/current/processor-decode-cef.html)
- [ ] [decode_csv_fields](https://www.elastic.co/guide/en/beats/filebeat/current/decode-csv-fields.html)
- [ ] [decode_duration](https://www.elastic.co/guide/en/beats/filebeat/current/decode-duration.html)
- [ ] [decode_xml](https://www.elastic.co/guide/en/beats/filebeat/current/decode-xml.html)
- [ ] [decode_xml_wineventlog](https://www.elastic.co/guide/en/beats/filebeat/current/decode-xml-wineventlog.html)
- [ ] [dissect](https://www.elastic.co/guide/en/beats/filebeat/current/dissect.html)
- [ ] [dns](https://www.elastic.co/guide/en/beats/filebeat/current/processor-dns.html)
- [ ] [extract_array](https://www.elastic.co/guide/en/beats/filebeat/current/extract-array.html)
- [ ] [fingerprint](https://www.elastic.co/guide/en/beats/filebeat/current/fingerprint.html)
- [ ] [move_fields](https://www.elastic.co/guide/en/beats/filebeat/current/move-fields.html)
- [ ] [parse_aws_vpc_flow_log](https://www.elastic.co/guide/en/beats/filebeat/current/processor-parse-aws-vpc-flow-log.html)
- [ ] [rate_limit](https://www.elastic.co/guide/en/beats/filebeat/current/rate-limit.html)
- [ ] [registered_domain](https://www.elastic.co/guide/en/beats/filebeat/current/processor-registered-domain.html)
- [ ] [syslog](https://www.elastic.co/guide/en/beats/filebeat/current/syslog.html)
- [ ] [timestamp](https://www.elastic.co/guide/en/beats/filebeat/current/processor-timestamp.html)
- [ ] [translate_sid](https://www.elastic.co/guide/en/beats/filebeat/current/processor-translate-sid.html)
- [ ] [urldecode](https://www.elastic.co/guide/en/beats/filebeat/current/urldecode.html)
- [x] [add_cloud_metadata](https://www.elastic.co/guide/en/beats/filebeat/current/add-cloud-metadata.html)
- [x] [include_fields](https://www.elastic.co/guide/en/beats/filebeat/current/include-fields.html)
- [x] [rename](https://www.elastic.co/guide/en/beats/filebeat/current/rename-fields.html)
- [x] [add_labels](https://www.elastic.co/guide/en/beats/filebeat/current/add-labels.html)
- [x] [copy_fields](https://www.elastic.co/guide/en/beats/filebeat/current/copy-fields.html)
- [x] [decode_json_fields](https://www.elastic.co/guide/en/beats/filebeat/current/decode-json-fields.html)
- [x] [detect_mime_type](https://www.elastic.co/guide/en/beats/filebeat/current/detect-mime-type.html)
- [x] [replace](https://www.elastic.co/guide/en/beats/filebeat/current/replace-fields.html)
- [x] [add_network_direction](https://www.elastic.co/guide/en/beats/filebeat/current/add-network-direction.html)
- [x] [decompress_gzip_field](https://www.elastic.co/guide/en/beats/filebeat/current/decompress-gzip-field.html)
- [x] [drop_fields](https://www.elastic.co/guide/en/beats/filebeat/current/drop-fields.html)
- [x] [truncate_fields](https://www.elastic.co/guide/en/beats/filebeat/current/truncate-fields.html)
- [x] [add_fields](https://www.elastic.co/guide/en/beats/filebeat/current/add-fields.html)
- [x] [add_tags](https://www.elastic.co/guide/en/beats/filebeat/current/add-tags.html)
- [x] [append](https://www.elastic.co/guide/en/beats/filebeat/current/append.html)
- [x] [decode_base64_field](https://www.elastic.co/guide/en/beats/filebeat/current/decode-base64-field.html)
- [x] [drop_event](https://www.elastic.co/guide/en/beats/filebeat/current/drop-event.html)
- [x] [script](https://www.elastic.co/guide/en/beats/filebeat/current/processor-script.html)

**BEFORE/AFTER benchmark results will be here once PR is finished**

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

I added unit tests and a benchmark.

### CODE REVIEW

File: libbeat/processors/processor_test.go [hunk 1]
```diff
@@ -566,3 +569,145 @@ func TestDropMissingFields(t *testing.T) {
...
+		{
+			"add_fields": mapstr.M{
+				"target": "added",
+				"fields": mapstr.M{
+					"field": "value",
+				},
+			},
+		},
+	}
+	processors := GetProcessors(t, cfg)
```
Comment (leehinman) [inline libbeat/processors/processor_test.go:589]: I think we need to test the full Create, Read, Update, and Delete scenarios.  I was thinking a javascript processor might be good for this.  We could test creating a new string, reading the new string, updating an existing string, and deleting an existing string, and then have the javascript processor either succeed or fail and then see if we have original event in failure case or modified event in success case.  We would want to do that for numbers, booleans, arrays and objects too.

Comment (rdner) [reply libbeat/processors/processor_test.go:589]: I think it won't be necessary once the processors list migrated to the new `EventEditor` that I'm working on, a lot is still going to change here.

EventEditor's tests will cover all operations on the `Event` including `Apply` and `Reset` of all changes made by processors. I added this test just to verify that `EventEditor` or `Event.Clone` are used and original nested maps don't change, nothing else.

**Comment by amitkanfer** [comment]
can't wait for this to land. will have a great positive impact on performance!

**Comment by rdner** [comment]
Closing in favour of https://github.com/elastic/beats/pull/36958

**Comment by amitkanfer** [comment]
> Closing in favour of https://github.com/elastic/beats/pull/36958

decided to first deliver a first set of processors?

**Pull request closed without merging**